src/bijectors/named_bijector.jl

```diff
@@ -0,0 +1,231 @@
+
+Wraps a `NamedTuple` of key -> `Bijector` pairs, implementing evaluation, inversion, etc.
+
+# Examples
+```julia-repl
+julia> using Bijectors: NamedBijector, Scale, Exp
+
+julia> b = NamedBijector((a = Scale(2.0), b = Exp()));
+
+julia> x = (a = 1., b = 0., c = 42.);
+
+julia> b(x)
+(a = 2.0, b = 1.0, c = 42.0)
+
+julia> (a = 2 * x.a, b = exp(x.b), c = x.c)
+(a = 2.0, b = 1.0, c = 42.0)
+```
+"""
+struct NamedBijector{names, Bs<:NamedTuple{names}} <: AbstractNamedBijector
+    bs::Bs
+end
+
+names_to_bijectors(b::NamedBijector) = b.bs
+
+@generated function (b::NamedBijector{names})(x::NamedTuple) where {names}
+    return :(merge(x, ($([:($n = b.bs.$n(x.$n)) for n in names]...), )))
+end
+
+@generated function Base.inv(b::NamedBijector{names}) where {names}
+    return :(NamedBijector(($([:($n = inv(b.bs.$n)) for n in names]...), )))
+end
+
+@generated function logabsdetjac(b::NamedBijector{names}, x::NamedTuple) where {names}
+    exprs = [:(logabsdetjac(b.bs.$n, x.$n)) for n in names]
+    return :(+($(exprs...)))
+end
+
+
+######################
+### `NamedInverse` ###
+######################
+"""
+    NamedInverse <: AbstractNamedBijector
+
+Represents the inverse of a `AbstractNamedBijector`, similarily to `Inverse` for `Bijector`.
+
+See also: [`Inverse`](@ref)
+"""
+struct NamedInverse{B<:AbstractNamedBijector} <: AbstractNamedBijector
+    orig::B
+end
+Base.inv(nb::AbstractNamedBijector) = NamedInverse(nb)
+Base.inv(ni::NamedInverse) = ni.orig
+
+logabsdetjac(ni::NamedInverse, y::NamedTuple) = -logabsdetjac(inv(ni), ni(y))
+
+##########################
+### `NamedComposition` ###
+##########################
+"""
+    NamedComposition <: AbstractNamedBijector
+
+Wraps a tuple of array of `AbstractNamedBijector` and implements their composition.
+
+This is very similar to `Composed` for `Bijector`, with the exception that we do not require
+the inputs to have the same "dimension", which in this case refers to the *symbols* for the
+`NamedTuple` that this takes as input.
+
+See also: [`Composed`](@ref)
+"""
+struct NamedComposition{Bs} <: AbstractNamedBijector
+    bs::Bs
+end
+
+# Essentially just copy-paste from impl of composition for 'standard' bijectors,
+# with minor changes here and there.
+composel(bs::AbstractNamedBijector...) = NamedComposition(bs)
+composer(bs::AbstractNamedBijector...) = NamedComposition(reverse(bs))
+∘(b1::AbstractNamedBijector, b2::AbstractNamedBijector) = composel(b2, b1)
+
+inv(ct::NamedComposition) = NamedComposition(reverse(map(inv, ct.bs)))
+
+function (cb::NamedComposition{<:AbstractArray{<:AbstractNamedBijector}})(x)
+    @assert length(cb.bs) > 0
+    res = cb.bs[1](x)
+    for b ∈ Base.Iterators.drop(cb.bs, 1)
+        res = b(res)
+    end
+
+    return res
+end
+
+@generated function (cb::NamedComposition{T})(x) where {T<:Tuple}
+    @assert length(T.parameters) > 0
+    expr = :(x)
+    for i in 1:length(T.parameters)
+        expr = :(cb.bs[$i]($expr))
+    end
+    return expr
+end
```

Comment:
I'm also wondering how much you actually gain by using a generated function here instead of also
```suggestion
(cb::NamedComposition{<:Tuple})(x) = fold(|>, cb.bs; init=x)
```

Reply:
Type-stability issue if output is different type than input, e.g. if we're using TrackedArray for parameters of the bijectors but not the input.

Reply:
For tuples (or at least small tuples) apparently the compiler can infer the output even if the output changes:
```julia
julia> f(x) = foldl(|>, x; init=1)

julia> @code_warntype f((x -> 2.0 * x, x -> x + 1))
Variables
  #self#::Core.Compiler.Const(f, false)
  x::Core.Compiler.Const((var"#31#33"(), var"#32#34"()), false)

Body::Float64
1 ─ %1 = (:init,)::Core.Compiler.Const((:init,), false)
│   %2 = Core.apply_type(Core.NamedTuple, %1)::Core.Compiler.Const(NamedTuple{(:init,),T} where T<:Tuple, false)
│   %3 = Core.tuple(1)::Core.Compiler.Const((1,), false)
│   %4 = (%2)(%3)::Core.Compiler.Const((init = 1,), false)
│   %5 = Core.kwfunc(Main.foldl)::Core.Compiler.Const(Base.var"#foldl##kw"(), false)
│   %6 = (%5)(%4, Main.foldl, Main.:|>, x)::Core.Compiler.Const(3.0, false)
└──      return %6
```

Reply:
Though maybe the example is too simple since the compiler even figures out the result.

Reply:
More realistic maybe:
```julia
julia> g(x, fs) = foldl(|>, fs; init=x)
g (generic function with 1 method)

julia> @code_warntype g(1, (x -> 2.0 * x, x -> x + 1))
Variables
  #self#::Core.Compiler.Const(g, false)
  x::Int64
  fs::Core.Compiler.Const((var"#35#37"(), var"#36#38"()), false)

Body::Float64
1 ─ %1 = (:init,)::Core.Compiler.Const((:init,), false)
│   %2 = Core.apply_type(Core.NamedTuple, %1)::Core.Compiler.Const(NamedTuple{(:init,),T} where T<:Tuple, false)
│   %3 = Core.tuple(x)::Tuple{Int64}
│   %4 = (%2)(%3)::NamedTuple{(:init,),Tuple{Int64}}
│   %5 = Core.kwfunc(Main.foldl)::Core.Compiler.Const(Base.var"#foldl##kw"(), false)
│   %6 = (%5)(%4, Main.foldl, Main.:|>, fs)::Float64
└──      return %6
```

Reply:
Uhh, I can't make it _not_ work, so you might be right :flushed: 
If so, that's waaaay better! I'm down for changing it :+1: